Show nicer error when a `@ConfigMapping` class exists in the `deployment` module and set to the `runtime` phase

### Description

While creating (see this PR for more context - https://github.com/quarkusio/quarkus-workshops/pull/600), I noticed that the following error is thrown when the `@ConfigMapping` phase is set to RUN_TIME:

```java
Caused by: java.lang.ClassNotFoundException: io.quarkus.workshop.superheroes.version.deployment.VersionConfig
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:525)
	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:533)
	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:481)
	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:533)
	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:481)
	at io.quarkus.runtime.configuration.AbstractConfigBuilder.withMapping(AbstractConfigBuilder.java:64)
	... 21 more

``` 

The solution was to move the `@ConfigMapping` to the `runtime` module, but the error message is a bit cryptic.

- Zulip thread: https://quarkusio.zulipchat.com/#narrow/stream/187030-users/topic/Workshop/near/471292983

### Implementation ideas

It would be nice if we could detect that somehow and display a nicer error message

--- a/core/processor/src/main/java/io/quarkus/annotation/processor/documentation/config/scanner/AbstractConfigListener.java
+++ b/core/processor/src/main/java/io/quarkus/annotation/processor/documentation/config/scanner/AbstractConfigListener.java
@@ -14,6 +14,8 @@ import io.quarkus.annotation.processor.documentation.config.discovery.DiscoveryC
 import io.quarkus.annotation.processor.documentation.config.discovery.EnumDefinition;
 import io.quarkus.annotation.processor.documentation.config.discovery.EnumDefinition.EnumConstant;
 import io.quarkus.annotation.processor.documentation.config.discovery.ResolvedType;
+import io.quarkus.annotation.processor.documentation.config.model.ConfigPhase;
+import io.quarkus.annotation.processor.documentation.config.model.ExtensionModule;
 import io.quarkus.annotation.processor.documentation.config.util.Types;
 import io.quarkus.annotation.processor.util.Config;
 import io.quarkus.annotation.processor.util.Utils;
@@ -64,6 +66,18 @@ public class AbstractConfigListener implements ConfigAnnotationListener {
         EnumDefinition enumDefinition = new EnumDefinition(enumTypeElement.getQualifiedName().toString(),
                 enumConstants);
         configCollector.addResolvedEnum(enumDefinition);
+    }
+
+    protected void validateRuntimeConfigOnDeploymentModules(ConfigPhase configPhase, TypeElement configRoot) {
+        if (configPhase.equals(ConfigPhase.RUN_TIME) || configPhase.equals(ConfigPhase.BUILD_AND_RUN_TIME_FIXED)) {
+            ExtensionModule.ExtensionModuleType type = config.getExtensionModule().type();
+            if (type.equals(ExtensionModule.ExtensionModuleType.DEPLOYMENT)) {
+                throw new IllegalStateException(String.format(
+                        "Error on %s: Configuration classes with ConfigPhase.RUN_TIME or " +
+                                "ConfigPhase.BUILD_AND_RUNTIME_FIXED phases, must reside in the respective module.",
+                        configRoot.getSimpleName().toString()));
+            }
+        }
     }
 
     protected void handleCommonPropertyAnnotations(DiscoveryConfigProperty.Builder builder,

--- a/core/processor/src/main/java/io/quarkus/annotation/processor/documentation/config/scanner/ConfigMappingListener.java
+++ b/core/processor/src/main/java/io/quarkus/annotation/processor/documentation/config/scanner/ConfigMappingListener.java
@@ -75,6 +75,8 @@ public class ConfigMappingListener extends AbstractConfigListener {
             }
         }
 
+        validateRuntimeConfigOnDeploymentModules(configPhase, configRoot);
+
         for (Map.Entry<? extends ExecutableElement, ? extends AnnotationValue> entry : configMappingAnnotion.getElementValues()
                 .entrySet()) {
             if ("prefix()".equals(entry.getKey().toString())) {

--- a/core/processor/src/main/java/io/quarkus/annotation/processor/documentation/config/scanner/LegacyConfigRootListener.java
+++ b/core/processor/src/main/java/io/quarkus/annotation/processor/documentation/config/scanner/LegacyConfigRootListener.java
@@ -79,6 +79,8 @@ public class LegacyConfigRootListener extends AbstractConfigListener {
             }
         }
 
+        validateRuntimeConfigOnDeploymentModules(configPhase, configRoot);
+
         String overriddenDocPrefix = null;
         if (configDocPrefixAnnotation != null) {
             for (Map.Entry<? extends ExecutableElement, ? extends AnnotationValue> entry : configDocPrefixAnnotation

